src/core/socialprovider.js

```diff
@@ -182,16 +182,21 @@ XMPPSocialProvider.prototype.connect = function(continuation) {
     // This may indicate a broken connection to XMPP.
     // TODO: handle this.
     this.logger.error('received unhandled close event', e);
+    if (this.status === 'ONLINE') {
+      this.ping_();
+    }
   }.bind(this));
   this.client.addListener('end', function(e) {
-    this.logger.error('received end event, status: ' + this.status, e);
-    if (this.status !== 'ONLINE') {
-      // Reject login promise.
+    if (this.status !== 'ONLINE' && this.client) {
```

Comment:
No need to delete this.client, as it will just be reset the next time login is called.  Checking this.client is just a way to check that logout has not yet been called.  If we don't have this check I'm not sure how to distinguish between an 'end' event caused by logout (not an error) and a more random 'end' event that is an error and requires calling logout or rejecting the login promise